### PR TITLE
pkg/esxi: API for booting ESXi from disks and CDROMs

### PR DESCRIPTION
**Pull request by hugelgupf**
tested: manually.

### CODE REVIEW

File: cmds/exp/esxiboot/esxiboot.go [hunk 1]
```diff
@@ -28,153 +29,57 @@
...
-
-	comment = '#'
-	sep     = "---"
-
-	uuidMagic = "VMWARE FAT16    "
-	uuidSize  = 32
-	partition = 5
+var (
+	cfg       = flag.StringP("config", "c", "", "ESXi config file")
+	cdrom     = flag.StringP("cdrom", "r", "", "ESXi CDROM boot device")
```
Comment (GanShun) [inline cmds/exp/esxiboot/esxiboot.go]: add -r to the comments at the top of the file

File: pkg/esxi/esxi.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+	n, err := f.ReadAt(data, int64(blockSize))
+	if err != nil {
+		return "", err
+	}
+	if n != uuidSize {
+		return "", io.ErrUnexpectedEOF
+	}
+
+	if magic := string(data[:len(uuidMagic)]); magic != uuidMagic {
+		return "", fmt.Errorf("bad uuid magic %q", magic)
```
Comment (GanShun) [inline pkg/esxi/esxi.go]: print the expected magic value too thanks.

File: pkg/esxi/esxi.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+	o.args += " " + uuid
+	return nil
+}
+
+func parse(configFile string) (options, error) {
+	dir := filepath.Dir(configFile)
+
+	f, err := os.Open(configFile)
+	if err != nil {
+		log.Fatal(err)
```
Comment (GanShun) [inline pkg/esxi/esxi.go]: don't use log.Fatal

File: pkg/esxi/esxi.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+		switch key {
+		case kernel:
+			opt.kernel = filepath.Join(dir, val)
+		case args:
+			opt.args = filepath.Join(dir, val)
+		case modules:
+			for _, tok := range strings.Split(val, sep) {
+				tok = strings.TrimSpace(tok)
+				entry := strings.Fields(tok)
+				entry[0] = filepath.Join(dir, entry[0])
```
Comment (GanShun) [inline pkg/esxi/esxi.go]: What is this line doing?

Comment (GanShun) [reply pkg/esxi/esxi.go]: is there an assumption that the first entry in each token is some kind of filename? It's not immediately clear to me what format this is trying to parse

Comment (hugelgupf) [reply pkg/esxi/esxi.go]: I'll comment and write a test for it, too

File: cmds/exp/esxiboot/esxiboot.go (hunk 1 above)
Comment (hugelgupf) [reply cmds/exp/esxiboot/esxiboot.go]: done

File: cmds/exp/esxiboot/esxiboot.go [hunk 5]
```diff
@@ -5,14 +5,16 @@
 // esxiboot executes ESXi kernel over the running kernel.
 //
 // Synopsis:
-//     esxiboot --config <config> [-d (--device)]
+//     esxiboot [-d --device -p --partition] [-c --config]
```
Comment (GanShun) [inline cmds/exp/esxiboot/esxiboot.go]: Fix synopsis too for -r

**Comment by hugelgupf** [comment]
ptal

**Pull request merged**